File: Problem-2-RootToLeafSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Pair:
    root = None
    summ = 0

    def __init__(self, root, summ):
        self.summ = summ
        self.root = root

    def getKey(self):
        return self.root

    def getValue(self):
        return self.summ


class Solution:
    result = 0

    def sumNumbers(self, root) :

        if not root:
            return 0

        stack = []
        currsumm = 0

        while root or stack:
            while root:
                currsumm = currsumm * 10 + root.val
                stack.append(Pair(root, currsumm))
                root = root.left

            p = stack.pop()

            root = p.getKey()
            currsumm = p.getValue()

            if not root.right and not root.left:
                self.result = self.result + currsumm

            root = root.right

        return self.result

File: Problem2_RootToLeafSum.cs
// Approach: recursively calculate pathNum and return it to parent when you are at leaf. At each non-leaf node add left path sum and right path sum to get total and return that.
// Time Complexity : O(n) traversing all the numbers
// Space Complexity: O(h) height of the tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int SumNumbers(TreeNode root) {
        return TraverseRec(root, 0);
    }
    
    public int TraverseRec(TreeNode node, int pathNum){
        if(node == null) return 0;
        
        pathNum = pathNum*10 + node.val;
        if(node.left == null && node.right == null){
            return pathNum;
        }
            
        var leftsum = TraverseRec(node.left, pathNum);
        var rightsum = TraverseRec(node.right, pathNum);
        
        return leftsum + rightsum ;
    }
}